### PR TITLE
Updated detection of Node and Japa test runners

### PR DESCRIPTION
**Pull request by fabnguess**
This PR aims to improve the detection of test runners used in projects, particularly for Node.js and Japa.

Currently the code only checks the test script in the package.json file. However, in many projects it is the test-only script that is responsible for running the tests. This prevents correct detection of Node's test runner.

Proposed changes:
Test-only script support: The code has been changed to also check the test-only script in the package.json file, in addition to the test script.

Japa verification: As per [Japa documentation](https://japa.dev/docs/installation), the presence of @japa/runner in devDependencies is now checked to identify projects using Japa as a test runner.

### CODE REVIEW

**Comment by PierreDemailly** [comment]
I think we can go further by checking if the test script contains another script instead of looking for `test-only` specifically.
For instance, we can imagine something like this:
`"test": "npm run test:ut && npm run test:ft"`
And many other patterns.

In this example, we can detect that `test:ut` and `test:ft` are used in the `test` script, so we'll check both of them, recursively (if `test:ut` contains another script then we'll check it, etc).

Another way would be checking all scripts that contains (or starts with) `test`, so if we keep the first example, we'll detect that both `test:ut` and `test:ft` contains or starts with `test` then we'll check these scripts.
This way is probably simpler, WDYT?

**Comment by fabnguess** [comment]
I @PierreDemailly, The second approach seems easier to implement, while the first one offers greater flexibility. So, which solution would you prefer us to focus on?

**Comment by PierreDemailly** [comment]
> I @PierreDemailly, The second approach seems easier to implement, while the first one offers greater flexibility. So, which solution would you prefer us to focus on?

You can go for the second one, but I also think to another edge case.
Imagine something like this:
```json
{
  "scripts": {
    "test": "node --test",
    "e2e": "playwright test"
  }
}
```
We would not detect `playwright`.

So here are 2 ideas:
- Keep the second idea, but add `e2e` to the keywords used to check scripts (check scripts that contains or starts with `test` *and* `e2e`. 
- Just check ALL scripts, this way we would even catch some unexpected scripts like:
```json
{
  "scripts": {
    "mocha": "mocha",
    "integration": "jest --config jest.integration.config.js"
  }
}
```

**Comment by fabnguess** [comment]
Well rated

**Comment by PierreDemailly** [comment]
@fabnguess Nevermind i'm dumb 🤣 
We detect these runners by checking if they are present in (dev) deps, script does not matter.

File: src/utils/getTestFrameworkName.ts [hunk 1]
```diff
@@ -2,12 +2,12 @@
 import type { PackageJSON } from "@nodesecure/npm-types";
 
 // CONSTANTS
-const kNodeTestRunnerRegex = /(?:node\s+--test|tsx)/;
+const kNodeTestRunnerRegex = /node\s+.*--test|tsx/;
```
Comment (PierreDemailly) [inline src/utils/getTestFrameworkName.ts]:
```suggestion
const kNodeTestRunnerRegex = /[node|tsx]\s+.*--test/;
```

File: src/utils/getTestFrameworkName.ts [hunk 2]
```diff
@@ -2,12 +2,12 @@
...
 
 export function getTestFrameworkName(packageJson: PackageJSON): string {
   const { devDependencies: deps = {}, scripts = {} } = packageJson;
 
-  if (kNodeTestRunnerRegex.test(scripts.test)) {
+  if (kNodeTestRunnerRegex.test(scripts.test) || kNodeTestRunnerRegex.test(scripts["test-only"]) || "tsx" in deps) {
```
Comment (PierreDemailly) [inline src/utils/getTestFrameworkName.ts]: `|| "tsx" in deps` is wrong.
`tsx` is a library used to run TS without building first. Here you'll detect `node:test` for any TS project that use `tsx` even if the projet don't use `node:test`. (same problem with the regexp)
```suggestion
  if (kNodeTestRunnerRegex.test(scripts.test) || kNodeTestRunnerRegex.test(scripts["test-only"])) {
```

**Comment by fabnguess** [comment]
Indeed, our concern is specifically related to the Node test runner. We could detect it differently, as typically, when using the Node test runner, dependencies like `glob` and `tsx` are present.

File: test/getTestFrameworkName.spec.ts [hunk 3]
```diff
@@ -56,12 +66,27 @@ describe("getTestFrameworkName()", () => {
...
+    it("Should return node:test for test-only script", () => {
+      const packageJson = {
+        name: "test",
+        version: "1.0.0",
+        scripts: {
+          "test-only": "glob -c \"node --loader=esmock --no-warnings --test-concurrency 1 --test\" \"test/**/*.test.js\""
+        }
+      };
+      assert.equal(getTestFrameworkName(packageJson), "node:test");
+    });
```
Comment (PierreDemailly) [inline test/getTestFrameworkName.spec.ts]: Can you add tests for TypeScript ? (with `tsx`)

Comment (fabnguess) [reply test/getTestFrameworkName.spec.ts]: Like this? ```it("Should return node:test if tsx is in dependencies", () => {
    const packageJson = {
      name: "test",
      version: "1.0.0",
      dependencies: {
        tsx: "^3.0.0"
      }
    };
    assert.equal(getTestFrameworkName(packageJson), "node:test");
  });```

Comment (PierreDemailly) [reply test/getTestFrameworkName.spec.ts]: No we don't care the dependency, like this:
```ts
it("Should return node:test for test script when using tsx", () => {
  const packageJson = {
    name: "test",
    version: "1.0.0",
    scripts: {
      test: "tsx --test"
    }
  };
  assert.equal(getTestFrameworkName(packageJson), "node:test");
});
```

Comment (fabnguess) [reply test/getTestFrameworkName.spec.ts]: ho ok

File: src/utils/getTestFrameworkName.ts [hunk 4]
```diff
@@ -2,12 +2,12 @@
...
 
 // CONSTANTS
-const kNodeTestRunnerRegex = /(?:node\s+--test|tsx)/;
+const kNodeTestRunnerRegex = /[node|tsx]\s+.*--test/;
 
 export function getTestFrameworkName(packageJson: PackageJSON): string {
   const { devDependencies: deps = {}, scripts = {} } = packageJson;
 
-  if (kNodeTestRunnerRegex.test(scripts.test)) {
+  if (kNodeTestRunnerRegex.test(scripts.test) || kNodeTestRunnerRegex.test(scripts["test-only"])) {
```
Comment (PierreDemailly) [inline src/utils/getTestFrameworkName.ts]: As discussed this can be improved, depends if you're down to go futher (cf [my first comment](https://github.com/dashlog/core/pull/96#issuecomment-2585308130)).
Otherwise I guess it's fine for our own projects and/or it can be improved in the future

**Review by PierreDemailly — APPROVED** [review]

File: src/utils/getTestFrameworkName.ts (hunk 4 above)
Comment (fabnguess) [reply src/utils/getTestFrameworkName.ts]: I would be tempted to try to do it as we discussed.

File: test/getTestFrameworkName.spec.ts
Comment (PierreDemailly) [inline test/getTestFrameworkName.spec.ts:1]: May we should add a recursive test ?
```ts
it("Should return node:test when a nested script uses node --test, recursively", () => {
  const packageJson = {
    name: "test",
    version: "1.0.0",
    scripts: {
      test: "npm run foo",
      foo: "npm run bar",
      bar: "node --test"
    }
  };
  assert.equal(getTestFrameworkName(packageJson), "node:test");
});
```

File: src/utils/getTestFrameworkName.ts [hunk 5]
```diff
@@ -2,12 +2,12 @@
...
 
 // CONSTANTS
-const kNodeTestRunnerRegex = /(?:node\s+--test|tsx)/;
+const kNodeTestRunnerRegex = /[node|tsx]\s+.*--test/;
 
 export function getTestFrameworkName(packageJson: PackageJSON): string {
   const { devDependencies: deps = {}, scripts = {} } = packageJson;
 
-  if (kNodeTestRunnerRegex.test(scripts.test)) {
+  if (kNodeTestRunnerRegex.test(scripts.test) || isUsingNodeTestRunner(scripts, "test")) {
```
Comment (PierreDemailly) [inline src/utils/getTestFrameworkName.ts]:
```suggestion
  if (isUsingNodeTestRunner(scripts, "test")) {
```
Since there is
```ts
if (scriptCommand && kNodeTestRunnerRegex.test(scriptCommand)) {
  return true;
}
```

**Review by PierreDemailly — APPROVED** [review]

**Review by fraxken — APPROVED** [review]
